pkg/middleware/stored_session.go

```diff
@@ -98,44 +103,53 @@ func (s *storedSessionLoader) getValidatedSession(rw http.ResponseWriter, req *h
 		return nil, nil
 	}
 
-	err = s.refreshSessionIfNeeded(rw, req, session)
+	if !s.isSessionRefreshNeeded(session) {
+		return session, nil
+	}
+
+	session, err = s.store.LoadWithLock(req)
```

Comment:
I think somewhere around here you need to handle nodes that don't receive the lock (e.g. NodeA gets the lock and performs the refresh, NodeB is too late, ticket is locked).

I think NodeB needs to block here until the lock is released. Then instead of getting the lock and trying to refresh itself, it pulls the session from storage (and uses it assuming NodeA just refreshed it).